### DISCUSSION TITLE
Merge methods per branch

### DISCUSSION THREAD

**Discussion by BenLorantfy** [community/community#8196] [Pull Requests]
It would be great if "allowed merge methods" were configurable per branch.  This would allow workflows such as:

- I only want to allow squash merging into a "dev" branch
- I only want to allow merge commits from dev -> main

I believe this is a common workflow to have a "development" or "release" branch.  Often we only want to allow squash merging into these branches to squash poor git histories.  

However, when we want to "release" by merging "dev" / "release" into "main" then we don't want to squash merge, otherwise all the granular history is lost.

**Comment by marcofugaro** [comment]
Same for me, my team is accidentally squashing commits to main.
The commits on dev are already squashed for feature, so all that info is lost.

**Comment by raurir** [comment]
This would be ideal for our team too.

**Comment by sam-w** [comment]
Absolutely, this would be an incredibly useful feature for us

**Comment by studoo** [comment]
This is exactly what our team wants to do. Makes it easier for everyone to not need to do a manual squash (which can be quite easily forgotten)

**Comment by nachozullo** [comment]
Same here, this is what our team need!

**Comment by KATT** [comment]
We had a new major of [tRPC](https://github.com/trpc/trpc) brewing for >6 months and it was a massive pain point for us to keep the branches up-to-date.

What we wanted was to:

- have the `main` branch to "Allow squash" only, &
- the `next-major`-branch to "Allow merge" only

**Comment by mskreativ** [comment]
+1

I need to define allowed and default merge method individually for some named branches.

**Comment by ChelseaMD** [comment]
+1

**Comment by chase-rg** [comment]
+1, this would be useful

**Comment by calvinzikakis-vizio** [comment]
+100 how is this not a feature still

**Comment by galamit86** [comment]
+1

**Comment by iamtodor** [comment]
We are lacking this feature. This is very handy to use when following the `git flow` strategy or `hotfix branch policy`

**Comment by calvinzikakis-vizio** [comment]
Duplicate of https://github.com/orgs/community/discussions/20583

**Reply by katylava** [reply]
I'd say the discussion is a duplicate of the issue, but everyone should definitely upvote both (with the upvote button, not a +1 comment).

Looks like the discussion is getting more traction so more upvotes there are probably more helpful.

This issue:
![Merge_methods_per_branch_·_community_·_Discussion__8196](https://github.com/community/community/assets/10575/cbfddc59-aa32-4da3-9edb-ac13a3a8f80a)

The discussion:
![Default_merge_strategy__per_org__repo_or_branch__·_community_·_Discussion__20583](https://github.com/community/community/assets/10575/893a24cf-e268-447e-ac08-bb1200e4ba20)

**Reply by ocean-orrick** [reply]
Default merge strategy per branch is a little different from enabling/disabling per branch. Both would be good. There is also this one with a large number of upvotes, which is asking for a subset of the feature requested here (disable squash and merge per branch) https://github.com/orgs/community/discussions/10809

**Reply by BenLorantfy** [reply]
I raised this issue 6 months before https://github.com/orgs/community/discussions/20583 so if anything https://github.com/orgs/community/discussions/20583 is a duplicate of this issue

I agree with upvoting both though

**Comment by huangmin-ms** [comment]
+1

**Comment by nam-truong-le** [comment]
+1

**Comment by mvettosi** [comment]
+1, my team really needs this!

**Comment by MaxBeauchemin** [comment]
+1

**Comment by bleadof** [comment]
+1

**Comment by malinindev** [comment]
+1

**Comment by JAnacleto** [comment]
If I understand correctly, this is something that I was looking for as well.

I was trying to configure the following in the 'main' branch rule set:
- Force merge commits when source branch is release/*
- Force squash merging when source branch is hotfix/*

This feature would really help to keep the branch history cleaner.

**Comment by ivanlynch** [comment]
+1

**Comment by benjaminclark177-lab** [comment]
That would definitely make branch strategies more flexible, especially for teams that separate development and release flows. A configurable system like [LemPlates](https://lemplates.com/) could help maintain clean histories in dev while preserving detailed commits when merging into main.

**Comment by DzmVasileusky** [comment]
Funny, that's already almost 5 years passed but we are still dealing merge conflicts and noise from master to develop PRs because people forgetting to manually switch merge strategy.
It this feature is so hard to implement in github's opinion?

**Reply by malinindev** [reply]
It looks like github has more important things to worry about :)
<img width="823" height="290" alt="image" src="https://github.com/user-attachments/assets/5254658e-feec-42d1-ac5d-d28df81c52e6" />

**Reply by DzmVasileusky** [reply]
Lol, yeah. It seems AI stroke them even more than Oracle guys
